airbyte-workers/src/main/java/io/airbyte/workers/WorkerUtils.java

```diff
@@ -53,17 +56,98 @@ public static void gentleClose(final Process process, final long timeout, final
       LOGGER.error("Exception while while waiting for process to finish", e);
     }
     if (process.isAlive()) {
-      LOGGER.warn("Process is taking too long to finish. Killing it");
-      process.destroy();
+      forceShutdown(process, 1, TimeUnit.MINUTES);
+    }
+  }
+
+  /**
+   * As long as the the heartbeatMonitor detects a heartbeat, the process will be allowed to continue.
+   * This method checks the heartbeat once every minute. Once there is no heartbeat detected, if the
+   * process has ended, then the method returns. If the process is still running it is given a grace
+   * period of the timeout arguments passed into the method. Once those expire the process is killed
+   * forcibly. If the process cannot be killed, this method will log that this is the case, but then
+   * returns.
+   *
+   * @param process - process to monitor.
+   * @param heartbeatMonitor - tracks if the heart is still beating for the given process.
+   * @param gracefulShutdownMagnitude - grace period to give the process to die after its heart stops
+   *        beating.
+   * @param gracefulShutdownTimeUnit - grace period to give the process to die after its heart stops
+   *        beating.
+   * @param checkHeartbeatMagnitude - frequency with which the heartbeat of the process is checked.
+   * @param checkHeartbeatTimeUnit - frequency with which the heartbeat of the process is checked.
+   * @param forcedShutdownMagnitude - amount of time to wait if a process needs to be destroyed
+   *        forcibly.
+   * @param forcedShutdownTimeUnit - amount of time to wait if a process needs to be destroyed
+   *        forcibly.
+   * @param forceShutdown
+   */
+  public static void gentleCloseWithHeartbeat(final Process process,
+                                              final HeartbeatMonitor heartbeatMonitor,
+                                              final long gracefulShutdownMagnitude,
+                                              final TimeUnit gracefulShutdownTimeUnit,
+                                              final long checkHeartbeatMagnitude,
+                                              final TimeUnit checkHeartbeatTimeUnit,
+                                              final long forcedShutdownMagnitude,
+                                              final TimeUnit forcedShutdownTimeUnit) {
```

Comment:
Any particular reason why we're using a combination of TimeUnit and magnitude here?  I'm not super familiar with Java, but it feels like this is a manual reinvention of [Duration.](https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html)

Reply:
good point. The pattern here you see in a lot of java method signatures. That being said, I don't have a good reason for it. I'll switch to duration, I think it will be less error prone.